zcash_transparent/src/zip48.rs

```diff
@@ -0,0 +1,450 @@
+    pub fn parse_key_info_expression<P: consensus::Parameters>(
+        s: &str,
+        params: &P,
+    ) -> Option<Self> {
+        match s.parse::<KeyExpression>().ok()?.into_parts() {
+            (Some(origin), descriptor::Key::Xpub { prefix, key, child })
+                if prefix == pub_prefix(params) && child.is_empty() =>
+            {
+                // Verify that this `KEY_INFO` expression is for ZIP 48.
+                match origin.derivation() {
+                    [purpose, coin_type, account, script_type]
+                        if purpose == &BIP_48_PURPOSE()
+                            && coin_type.index() == params.coin_type()
+                            && coin_type.is_hardened()
+                            && account.is_hardened()
+                            && script_type == &ZCASH_P2SH_SCRIPT_TYPE() =>
+                    {
+                        Some(Self { origin, key })
+                    }
+                    _ => None,
+                }
+            }
+            _ => None,
+        }
+    }
+
+    /// Encodes this public key as a [BIP 388 `KEY_INFO` expression].
+    ///
+    /// [BIP 388 `KEY_INFO` expression]: https://github.com/bitcoin/bips/blob/master/bip-0388.mediawiki#key-information-vector
+    pub fn key_info_expression<P: consensus::Parameters>(&self, params: &P) -> String {
+        self.key_expression_inner(pub_prefix(params), vec![])
+            .to_string()
+    }
+
+    /// Produces a [`KeyExpression`] from this public key corresponding to a specific
+    /// address.
+    pub fn key_expression_for_address(
+        &self,
+        prefix: Prefix,
+        scope: zip32::Scope,
+        address_index: NonHardenedChildIndex,
+    ) -> KeyExpression {
+        self.key_expression_inner(
+            prefix,
+            vec![
+                TransparentKeyScope::from(scope).into(),
+                address_index.into(),
+            ],
+        )
+    }
+
+    fn key_expression_inner(
+        &self,
+        prefix: Prefix,
+        child: Vec<bip32::ChildNumber>,
+    ) -> KeyExpression {
+        KeyExpression::from_xpub(Some(self.origin.clone()), prefix, self.key.clone(), child)
+            .expect("correct by construction")
+    }
+}
+
+/// A [ZIP 48] P2SH multisig full viewing key.
+///
+/// This provides the necessary derivation capability to view all funds controlled by a
+/// P2SH multisig account.
+///
+/// [ZIP 48]: https://zips.z.cash/zip-0048
+pub struct FullViewingKey {
+    threshold: u8,
+    key_info: Vec<AccountPubKey>,
+}
+
+impl FullViewingKey {
+    /// Constructs a full viewing key for a standard ZIP 48 account.
+    ///
+    /// This uses the standard wallet descriptor template. For example:
+    /// - `threshold == 2`
+    /// - `key_info.len() == 3`
+    /// - Wallet descriptor template: `"sh(sortedmulti(2,@0/**,@1/**,@2/**))"`
+    pub fn standard(
+        threshold: u8,
+        key_info: Vec<AccountPubKey>,
```

Comment:
```suggestion
        key_info: &[AccountPubKey],
```

Reply:
We take ownership of `key_info` here. If we took a slice, we'd then need to clone it.